source/common/uma_pools/usm_allocator.cpp

```diff
@@ -0,0 +1,866 @@
+  std::unordered_multimap<void *, Slab &> KnownSlabs;
+  std::shared_timed_mutex KnownSlabsMapLock;
+
+  // Handle to the memory allocation routine
+  std::unique_ptr<SystemMemory> MemHandle;
+
+  // Store as unique_ptrs since Bucket is not Movable(because of std::mutex)
+  std::vector<std::unique_ptr<Bucket>> Buckets;
+
+  // Configuration for this instance
+  USMAllocatorParameters params;
+
+public:
+  USMAllocImpl(std::unique_ptr<SystemMemory> SystemMemHandle,
+               USMAllocatorParameters params)
+      : MemHandle{std::move(SystemMemHandle)}, params(params) {
+
+    // Generate buckets sized such as: 64, 96, 128, 192, ..., CutOff.
+    // Powers of 2 and the value halfway between the powers of 2.
+    auto Size1 = params.MinBucketSize;
+    auto Size2 = Size1 + Size1 / 2;
+    for (; Size2 < CutOff; Size1 *= 2, Size2 *= 2) {
+      Buckets.push_back(std::make_unique<Bucket>(Size1, *this));
+      Buckets.push_back(std::make_unique<Bucket>(Size2, *this));
+    }
+    Buckets.push_back(std::make_unique<Bucket>(CutOff, *this));
+  }
+
+  void *allocate(size_t Size, size_t Alignment, bool &FromPool);
+  void *allocate(size_t Size, bool &FromPool);
+  void deallocate(void *Ptr, bool &ToPool, bool OwnZeMemHandle);
+
+  SystemMemory &getMemHandle() { return *MemHandle; }
+
+  std::shared_timed_mutex &getKnownSlabsMapLock() { return KnownSlabsMapLock; }
+  std::unordered_multimap<void *, Slab &> &getKnownSlabs() {
+    return KnownSlabs;
+  }
+
+  size_t SlabMinSize() { return params.SlabMinSize; };
+
+  USMAllocatorParameters &getParams() { return params; }
+
+  void printStats(bool &TitlePrinted, size_t &HighBucketSize,
+                  size_t &HighPeakSlabsInUse, const std::string &Label);
+
+private:
+  Bucket &findBucket(size_t Size);
+};
+
+bool operator==(const Slab &Lhs, const Slab &Rhs) {
+  return Lhs.getPtr() == Rhs.getPtr();
+}
+
+std::ostream &operator<<(std::ostream &Os, const Slab &Slab) {
+  Os << "Slab<" << Slab.getPtr() << ", " << Slab.getEnd() << ", "
+     << Slab.getBucket().getSize() << ">";
+  return Os;
+}
+
+Slab::Slab(Bucket &Bkt)
+    : // In case bucket size is not a multiple of SlabMinSize, we would have
+      // some padding at the end of the slab.
+      Chunks(Bkt.SlabMinSize() / Bkt.getSize()), NumAllocated{0}, bucket(Bkt),
+      SlabListIter{}, FirstFreeChunkIdx{0} {
+  auto SlabSize = Bkt.SlabAllocSize();
+  MemPtr = Bkt.getMemHandle().allocate(SlabSize);
+  regSlab(*this);
+}
+
+Slab::~Slab() {
+  unregSlab(*this);
+  bucket.getMemHandle().deallocate(MemPtr, true /* OwnZeMemHandle */);
+}
+
+// Return the index of the first available chunk, -1 otherwize
+size_t Slab::FindFirstAvailableChunkIdx() const {
+  // Use the first free chunk index as a hint for the search.
+  auto It = std::find_if(Chunks.begin() + FirstFreeChunkIdx, Chunks.end(),
+                         [](auto x) { return !x; });
+  if (It != Chunks.end()) {
+    return It - Chunks.begin();
+  }
+
+  return static_cast<size_t>(-1);
```

Comment:
Casting a negative int to the unsigned type.

Reply:
Done.